### PR TITLE
feat: allow custom os images from external subs via Shared Image Galleries

### PR DESCRIPTION
**Pull request by alexeldeib**
**Reason for Change**:

Allow users to create an image, push it to a shared gallery, replicate it to the rest of Azure, and deploy templates based off that.

**Issue Fixed**:

#721 

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

TODO:
- [ ] Validation on inputs for imageReference
- [x] Test test test

### CODE REVIEW

**Comment by alexeldeib** [comment]
Waiting on my OS image to propagate to Azure regions so I can do a test run with this. 

Test cases:
- [ ] Should work on master VMAS
- [ ] Should work on master VMSS
- [ ] Should work on worker VMAS
- [ ] Should work on worker VMSS
- [ ] don't regress existing behavior -- e.g. if unspecificed, should default to normal image.
- [ ] using a location where image does not exist should fail fast (?)

TODO
- [ ] using a location where image does not exist should fail fast (?)

**Comment by coreywagehoft** [comment]
@alexeldeib are there plans on finishing this feature?  This would be extremely helpful to a project I am working on right now.

**Comment by alexeldeib** [comment]
@coreywagehoft Apologies for letting this slip! This was more or less functional at the time, I had set up a few test subscriptions to test various permutations but replication took a looong time so I was only able to test simple scenarios. 

I need a rebase and some unit tests, plus maybe a non-PR blocking E2E test to complete this. @CecileRobertMichon @jackfrancis does that still sound reasonable? Sorry for letting this one fall by the wayside.

I'm traveling abroad until this weekend, but should be able to clean this up for review in the next ~week hopefully.

**Comment by CecileRobertMichon** [comment]
@alexeldeib sounds great, I look forward to this going in.

**Comment by alexeldeib** [comment]
Quick update: I've been testing this. Unfortunately, something with the custom images causes takes an extremely long time (>30 minutes of VM stuck in "Creating"). I was able to get ssh access to one machine briefly and it did have k8s manifests dowloaded, but did not have kubelet running yet. It booted me off shortly after. Trying to figure out if it's something in my setup, or with shared images. The image I created was extremely vanilla. 

```bash
az group create -n generalvm -l westus2
az vm create -g generalvm -n instance --image UbuntuLTS --ssh-key-values blah
ssh user@IPADDRESS
> sudo waagent -deprovision
> exit
az  vm deallocate -g generalvm -n instance
az vm generalize -g generalvm -n instance
az image create -g generalvm -n instanceImage --source instance
az sig create -g generalvm --gallery-name siggallery
az sig image-definition create -g generalvm --gallery-name siggallery --gallery-image-definition instancevm --publisher AceCo --offer Standard --sku alpha.0 --os-Type Linux
az sig image-version create -g generalvm --gallery-name siggallery --gallery-image-definition instancevm --gallery-image-version 0.0.1 --target-regions "westus2" --replication-count 5 --managed-image <<ResourceID from generalize image>>
```

Results: 
image
image
image

I'm definitely using a generalized image, not sure what i'm missing. I've occasionally gotten a master node mostly working, but something is clearly weird here

@CecileRobertMichon do you know anyone on the compute side closer to Shared Images? Might try to reach out to them to see what's up here. 

I'll push the rebased bits, maybe someone will see something obvious I missed.

File: pkg/engine/virtualmachines.go [hunk 1]
```diff
@@ -168,8 +167,12 @@ func CreateMasterVM(cs *api.ContainerService) VirtualMachineARM {
 		}
 	}
 	imgReference := &compute.ImageReference{}
-	if useMasterCustomImage {
-		imgReference.ID = to.StringPtr("[resourceId(parameters('osImageResourceGroup'), 'Microsoft.Compute/images', parameters('osImageName'))]")
+	if cs.Properties.MasterProfile.HasImageRef() {
+		if cs.Properties.MasterProfile.HasImageGallery() {
+			imgReference.ID = to.StringPtr(fmt.Sprintf("[concat('/subscriptions/', '%s', '/resourceGroups/', parameters('osImageResourceGroup'), '/providers/Microsoft.Compute/galleries/', '%s', '/images/', parameters('osImageName'), '/versions/', '%s')]", imageRef.SubscriptionID, imageRef.Gallery, imageRef.Version))
```
Comment (alexeldeib) [inline pkg/engine/virtualmachines.go]: ARM didn't like Resource IDs with `Microsoft.Compute/galleries/images/versions`

**Comment by michaeldisalvo** [comment]
Hi @alexeldeib, do you have any idea when folks from compute may be able to assist/troubleshoot?  Currently working on the project @coreywagehoft mentioned and we're a bit blocked waiting for this support.  Any info would be helpful so that we can sequence/update remaining work.  Thank you!

**Comment by alexeldeib** [comment]
@michaeldisalvo we've started offline discussion to debug with the SIG folks but no fix yet. As far as I can tell this implementation should work; if you're using Shared Image Galleries and have a chance to test, maybe the problem is simply with my image setup and you will have better luck. It would certainly provide another data point for us to investigate.

I'll update back if we make progress offline.

**Comment by alexeldeib** [comment]
Quick update, seems the issue is only around initial replication. Despite the replication being complete, it took some time until I was able to provision VMs as usual. 

Hit a few issues with CSE off of master, but I think this will be working in short order (pulling in Cecile's latest fixes and doing some validation).

Needs some tests.

**Comment by alexeldeib** [comment]
Unfortunately still intermittently hitting issues with VM create, but as far as I can tell it's unrelated to the code changes here. Trying with different base OS images to narrow down possible issues.

If anyone is feeling bold and has shared images they want to use with this, I'd appreciate more anecdata 😄

**Comment by coreywagehoft** [comment]
@alexeldeib We have seen in rare cases the same failure from time to time, even with no configuration changes.  A few weeks ago I was getting failures for several hours deploying a cluster over and over, and the next day I would not have issues.

**Comment by alexeldeib** [comment]
Great news -- we have liftoff :tada: Will clean up and rebase soon. I'm happy this is working!

```
ace@ace-x1:~/go/src/github.com/Azure/aks-engine$ kubectl --kubeconfig _output/banana4/kubeconfig/kubeconfig.westus2.json get node
NAME                                 STATUS   ROLES    AGE   VERSION
k8s-agentpool1-26857177-vmss000000   Ready    agent    48s   v1.12.9
k8s-master-26857177-0                Ready    master   44s   v1.12.9
ace@ace-x1:~/go/src/github.com/Azure/aks-engine$ 
```

**Comment by alexeldeib** [comment]
I'm working through my queue tonight before heading for a week of vacation tomorrow morning, right now I don't expect to be able to finish this before leaving, unfortunately. My apologies for the additional delay guys, this is one of the top items on my list to finish.

**Comment by justinsmits** [comment]
Hey @alexeldeib , can you please provide an update on when you expect to complete this, and if possible let us know when there will be a new version of the engine with this support? This issue is blocking a critical project within our business.

File: pkg/engine/virtualmachinescalesets.go [hunk 2]
```diff
@@ -577,11 +580,17 @@ func CreateAgentVMSS(cs *api.ContainerService, profile *api.AgentPoolProfile) Vi
 		}
 		vmssStorageProfile.DataDisks = getVMSSDataDisks(profile)
 	} else {
-		imageRef := profile.ImageRef
-		useAgentCustomImage := imageRef != nil && len(imageRef.Name) > 0 && len(imageRef.ResourceGroup) > 0
-		if useAgentCustomImage {
-			vmssStorageProfile.ImageReference = &compute.ImageReference{
-				ID: to.StringPtr(fmt.Sprintf("[resourceId(variables('%[1]sosImageResourceGroup'), 'Microsoft.Compute/images', variables('%[1]sosImageName'))]", profile.Name)),
+		if profile.HasImageRef() {
+			imageRef := profile.ImageRef
```
Comment (CecileRobertMichon) [inline pkg/engine/virtualmachinescalesets.go]: these new lines don't have UT coverage, see https://github.com/Azure/aks-engine/blob/master/pkg/engine/virtualmachinescalesets_test.go#L701 for an example of a VMSS unit test.

File: pkg/engine/virtualmachines.go [hunk 3]
```diff
@@ -475,10 +478,15 @@ func createAgentAvailabilitySetVM(cs *api.ContainerService, profile *api.AgentPo
 
 	} else {
 		imageRef := profile.ImageRef
-		useAgentCustomImage := imageRef != nil && len(imageRef.Name) > 0 && len(imageRef.ResourceGroup) > 0
-		if useAgentCustomImage {
-			storageProfile.ImageReference = &compute.ImageReference{
-				ID: to.StringPtr(fmt.Sprintf("[resourceId(variables('%[1]sosImageResourceGroup'), 'Microsoft.Compute/images', variables('%[1]sosImageName'))]", profile.Name)),
+		if profile.HasImageRef() {
+			if profile.HasImageGallery() {
```
Comment (CecileRobertMichon) [inline pkg/engine/virtualmachines.go]: ditto here for unit test coverage

File: pkg/api/types.go [hunk 4]
```diff
@@ -1305,6 +1320,18 @@ func (m *MasterProfile) GetCosmosEndPointURI() string {
...
+// HasImageRef returns true if the customer brought os image
+func (a *AgentPoolProfile) HasImageRef() bool {
+	imageRef := a.ImageRef
+	return imageRef != nil && len(imageRef.Name) > 0 && len(imageRef.ResourceGroup) > 0
+}
+
+// HasImageGallery returns true if the customer brought os image from Shared Image Gallery
+func (a *AgentPoolProfile) HasImageGallery() bool {
+	imageRef := a.ImageRef
+	return imageRef != nil && len(imageRef.SubscriptionID) > 0 && len(imageRef.Gallery) > 0 && len(imageRef.Version) > 0
```
Comment (CecileRobertMichon) [inline pkg/api/types.go]: this isn't covered by unit tests

File: pkg/api/types.go [hunk 5]
```diff
@@ -1188,6 +1191,18 @@ func (p *Properties) GetMasterFQDN() string {
...
+// HasImageRef returns true if the customer brought os image
+func (m *MasterProfile) HasImageRef() bool {
+	imageRef := m.ImageRef
+	return imageRef != nil && len(imageRef.Name) > 0 && len(imageRef.ResourceGroup) > 0
+}
+
+// HasImageGallery returns true if the customer brought os image from Shared Image Gallery
+func (m *MasterProfile) HasImageGallery() bool {
+	imageRef := m.ImageRef
+	return imageRef != nil && len(imageRef.SubscriptionID) > 0 && len(imageRef.Gallery) > 0 && len(imageRef.Version) > 0
```
Comment (CecileRobertMichon) [inline pkg/api/types.go]: this isn't covered by unit tests

File: pkg/api/converterfromapi.go [hunk 6]
```diff
@@ -1049,6 +1052,9 @@ func convertAgentPoolProfileToVLabs(api *AgentPoolProfile, p *vlabs.AgentPoolPro
 		p.ImageRef = &vlabs.ImageReference{}
 		p.ImageRef.Name = api.ImageRef.Name
```
Comment (CecileRobertMichon) [inline pkg/api/converterfromapi.go]: this isn't covered by unit tests

File: pkg/api/convertertoapi.go [hunk 7]
```diff
@@ -1065,6 +1068,9 @@ func convertVLabsAgentPoolProfile(vlabs *vlabs.AgentPoolProfile, api *AgentPoolP
 		api.ImageRef = &ImageReference{}
 		api.ImageRef.Name = vlabs.ImageRef.Name
 		api.ImageRef.ResourceGroup = vlabs.ImageRef.ResourceGroup
```
Comment (CecileRobertMichon) [inline pkg/api/convertertoapi.go]: this also isn't covered by unit tests

File: examples/custom-shared-image.json [hunk 8]
```diff
@@ -0,0 +1,50 @@
+{
+    "apiVersion": "vlabs",
```
Comment (CecileRobertMichon) [inline examples/custom-shared-image.json]: can you please add some docs for the feature? 

- apimodel definition should go in https://github.com/Azure/aks-engine/blob/master/docs/topics/clusterdefinitions.md#L513
- feature documentation can go in:
https://github.com/Azure/aks-engine/blob/master/docs/topics/features.md

**Comment by alexeldeib** [comment]
@CecileRobertMichon I'm flying back from vacation in a few hours, I'll be able to clean this up tomorrow. Thanks for reviewing 👍

**Comment by CecileRobertMichon** [comment]
@alexeldeib sounds great, don't hesitate to reach out if you need help moving this forward. I added a bunch of "missing unit tests" comments. You can see the coverage diff at https://codecov.io/gh/Azure/aks-engine/pull/913/diff if that's helpful.

**Comment by CecileRobertMichon** [comment]
@justinsmits we are planning on cutting v0.38.0 early next week so if the PR makes it to master before then it will be included in that release.

**Comment by justinsmits** [comment]
Thanks, @CecileRobertMichon and @alexeldeib! Do you think there is a high or low probability that it will make it in to the release next week? I need to report this up to senior management, and I am sure they will want to know how likely we are to get this feature implemented. Thanks again!

**Comment by alexeldeib** [comment]
@CecileRobertMichon pending successful tests, PTAL

File: pkg/api/converterfromapi.go (hunk 6 above)
Comment (alexeldeib) [reply pkg/api/converterfromapi.go]: I added an agent pool to cover this in to/from conversion test, vlabs version only.

File: pkg/api/convertertoapi.go (hunk 7 above)
Comment (alexeldeib) [reply pkg/api/convertertoapi.go]: same comment as convertfrom file

File: pkg/api/types.go (hunk 5 above)
Comment (alexeldeib) [reply pkg/api/types.go]: tested helpers in VMSS and VMAS

**Comment by alexeldeib** [comment]
Codecov is failing because the test coverage was low prior to the PR, I think I only increased the net and % of diff is above target in all the files. LMK if I missed anything there.

**Comment by alexeldeib** [comment]
Looks like transient time sync failure on a couple e2e? Not sure if I need to fix, LMK

**Comment by CecileRobertMichon** [comment]
transient time sync failure with 18.04 is a [known issue](https://github.com/Azure/aks-engine/issues/1306), I restarted the job.

File: pkg/api/types.go [hunk 9]
```diff
@@ -1196,6 +1199,16 @@ func (p *Properties) GetMasterFQDN() string {
 	return p.MasterProfile.FQDN
 }
 
+// HasImageRef returns true if the customer brought os image
+func (m *MasterProfile) HasImageRef() bool {
+	return m.ImageRef != nil && len(m.ImageRef.Name) > 0 && len(m.ImageRef.ResourceGroup) > 0
+}
+
+// HasImageGallery returns true if the customer brought os image from Shared Image Gallery
+func (m *MasterProfile) HasImageGallery() bool {
```
Comment (CecileRobertMichon) [inline pkg/api/types.go]: missing unit test here

File: pkg/api/types.go (hunk 4 above)
Comment (CecileRobertMichon) [reply pkg/api/types.go]: still missing

File: pkg/engine/virtualmachines.go [hunk 10]
```diff
@@ -149,7 +149,6 @@ func CreateMasterVM(cs *api.ContainerService) VirtualMachineARM {
 
 	storageProfile := &compute.StorageProfile{}
 	imageRef := cs.Properties.MasterProfile.ImageRef
-	useMasterCustomImage := imageRef != nil && len(imageRef.Name) > 0 && len(imageRef.ResourceGroup) > 0
```
Comment (CecileRobertMichon) [inline pkg/engine/virtualmachines.go]: thanks for making this a func

File: pkg/api/types_test.go [hunk 11]
```diff
@@ -1005,6 +1005,130 @@ func TestPropertiesIsHostedMasterProfile(t *testing.T) {
...
+			expectedRef:     false,
+			expectedGallery: false,
+		},
+	}
+
+	for _, c := range cases {
+		c := c
+		t.Run(c.name, func(t *testing.T) {
+			t.Parallel()
+			if c.p.AgentPoolProfiles[0].HasImageRef() != c.expectedRef && c.p.AgentPoolProfiles[0].HasImageGallery() != c.expectedGallery {
```
Comment (alexeldeib) [inline pkg/api/types_test.go]: @CecileRobertMichon isn't this testing the thing that codecov is still saying is untested? Am I testing the wrong type, or doing something dumb here?

Comment (CecileRobertMichon) [reply pkg/api/types_test.go]: I think the `&&` was the issue, your latest commit should fix it

**Comment by alexeldeib** [comment]
Think that was it because it was shortcircuiting the conditional, nvm.

**Comment by alexeldeib** [comment]
failing mostly on cluster RG deletion, not sure about windows

**Comment by alexeldeib** [comment]
@justinsmits @michaeldisalvo @coreywagehoft FYI 🎉

**Comment by justinsmits** [comment]
It's Christmas in July! Thank you!

**Comment by michaeldisalvo** [comment]
@alexeldeib Thank you for turning this around!  Is this feature intended to work with Windows agent profiles?  We're deploying clusters with the overrides set for `imageReference` on the agent pool with values for `windowsProfile` only containing the admin u/p combination however our test clusters are still coming up with the default image reference (Publisher: MicrosoftWindowsServer, Offer: WindowsServer, SKU: Datacenter-Core-1809-with-Containers-smalldisk, Version: 17763.557.20190604)

**Comment by alexeldeib** [comment]
@michaeldisalvo This only added linux support. I'm not very familiar with building custom Windows images or running Windows nodes with Kubernetes, so I'm not able to help validate and fix this properly. The changes are quite short, you can see the main ones below where windows could be added. The validation + any debug would be the trickier part.

I'd have to defer to the AKS engine team for more detailed answer (I work in Azure, but elsewhere; I am an AKS engine end-user). 

fyi @CecileRobertMichon 

The functionality would go here: 
- VM: https://github.com/Azure/aks-engine/blob/3481e200619b8b45d52fc21b1354cd69a9e7c609/pkg/engine/virtualmachines.go#L461-L500
- VMSS: https://github.com/Azure/aks-engine/blob/3481e200619b8b45d52fc21b1354cd69a9e7c609/pkg/engine/virtualmachinescalesets.go#L574-L604

**Comment by alexeldeib** [comment]
Sorry for the lack of clarity, I should PR a doc update with that info until/unless Windows does the same.

**Pull request merged**